### PR TITLE
Updated contributing.md and the style guide

### PR DESCRIPTION
**Pull request by ravi-mosaicml**
Added information on docstrings, conditional imports, and running tests.

### CODE REVIEW

**Comment by hanlint** [comment]
Should we merge style guide and contributing together?

**Comment by ravi-mosaicml** [comment]
> Should we merge style guide and contributing together?

I was thinking the same thing haha...let's do it then :)

**Comment by kobindra** [comment]
Looks like I'm a few steps behind -- review no longer needed, correct?

**Comment by ravi-mosaicml** [comment]
> Looks like I'm a few steps behind -- review no longer needed, correct? In other words, you're going to merge the files and open a new PR... or update this PR with the files merged, and then review will be needed?

I updated the PR with the requested changes....if you could give it a review that would be awesome!

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -4,4 +4,415 @@ Thanks for considering contributing to Composer!
...
+The following guidelines apply to documentation.
+1.  Each function that needs a docstring must have its input arguments and return statement (if not None) annotated.
+1.  Each argument annotation should include the type. If the argument has a default value, the type annotation should
+    specify "optional", and the docstring should say the default value. There should only be one line for the "Returns:"
+    section, even if a tuple is returned. For example:
+
+    ```python
+
+    def foo(bar: int):
+        """Foo.
```
Comment (dblalock) [inline CONTRIBUTING.md:313]: Can these one-liners be examples of using imperative rather than descriptive style, at least if we care about that? E.g.:
`"""Add 5 to the input"""`
as opposed to:
`"""Adds 5 to the input"""`

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -4,4 +4,415 @@ Thanks for considering contributing to Composer!
...
+                my_function(x)
+        """
+        ...
+    ```
+
+    To check doctests, run:
+    
+    ```bash
+    cd docs && make clean && make html && make doctest
+    ```
```
Comment (dblalock) [inline CONTRIBUTING.md:408]: already mentioned in section on running the tests. also we should probably have `make doctest` in the root directory run the doctests in one command, but that's a separate issue

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -4,4 +4,415 @@ Thanks for considering contributing to Composer!
...
+
+## 8. Hparams and Config Classes
+
+Each class that could be initialized by the hparams will need its configuration variables defined in
+[yahp](https://github.com/mosaicml/yahp) dataclass.
```
Comment (dblalock) [inline CONTRIBUTING.md:413]: Not sure what this sentence means. What does it mean to be "initialized by the hparams"? How do I know if a class I write could be initialized in this way? also missing "a" before "yahp".

Also, will future code have an associated yahp dataclass? If not, can maybe just rm this section

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -4,4 +4,415 @@ Thanks for considering contributing to Composer!
...
+
+Guidelines:
+* The fields in the dataclass should roughly follow the init signature of the class being constructed
+* `initialize_object(self)` should take any parameters needed to construct the class. It acceptable to
+take complex types or other hparam dataclasses, as required, to initialize the object.
```
Comment (dblalock) [inline CONTRIBUTING.md:418]: "It acceptable" -> "It is acceptable"

**Review by dblalock — COMMENTED** [review]
Just a bunch of little comments. Only a couple get at questions regarding the actual content.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -4,4 +4,415 @@ Thanks for considering contributing to Composer!
...
+        Args:
+            bar (int, optional): Optional bar. (default: ``42``)
+        """
+        ...
+    
+    def foo3(bar: Optional[int] = None):
+        """Foo3.
+
+        Args:
+            bar (int, optional): Optional bar. (default: ``None``)
```
[the remark this answers is not in the document]
Comment (hanlint) [reply CONTRIBUTING.md:332]: I may have been mistaken when I effected that in #662 as our Trainer does (default: `42`).

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -4,4 +4,415 @@ Thanks for considering contributing to Composer!
...
+    def foo(x: Union[int, None]):
+        assert x is not None, "x should never be None"
+        return x + 5  # valid
+    ```
+
+1. For variables where it is impossible for PyRight to infer the correct type, use
+[cast](https://docs.python.org/3/library/typing.html#typing.cast).
+1. As a last resort, add a `# type: ignore` comment to the line where PyRight emits an error.
+Immediately following this statement, paste in the error emitted by PyRight,
+so other contributors will know why this error was silenced.
```
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply CONTRIBUTING.md:154]: I prefer to include the full error message. Does Pyright support the bracket notation?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -4,4 +4,415 @@ Thanks for considering contributing to Composer!
...
+* Public methods in public classes (e.g. `composer.trainer.trainer.Trainer.fit`)
+* Public modules (e.g. `composer.trainer.trainer`)
+
+The following rules apply to public APIs:
+1. All public APIs must have a docstring (see the Documentation section below)
+1. All parameters must have type annotations.
+1. Parameters should be flattened, rather than nested. While this violates the software engineering practice of
+    being DRY (don't repeat yourself), it simplifies the user API by making all parameters visible.
+1. To minimize user imports, parameters should support native PyTorch types or Python primitives whenever possible.
+
```
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply CONTRIBUTING.md:172]: Agree with mapping over dict (though note that mapping is immutable, so a mutablemapping might need to be used sometimes). Re sequences, that will depend on #677.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -4,4 +4,415 @@ Thanks for considering contributing to Composer!
...
+## 7. Documentation
+
+Composer uses [Google Style Docstrings](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).
+
+The following guidelines apply to documentation.
+1.  Each function that needs a docstring must have its input arguments and return statement (if not None) annotated.
+1.  Each argument annotation should include the type. If the argument has a default value, the type annotation should
+    specify "optional", and the docstring should say the default value. There should only be one line for the "Returns:"
+    section, even if a tuple is returned. For example:
```
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply CONTRIBUTING.md:308]: Ah good catch; didn't realize google style supported this. Agree here; let's do it this way then.

**Comment by hanlint** [comment]
@dblalock I updated the files as discussed offline.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -5,3 +5,71 @@ Thanks for considering contributing to Composer!
 Issues tagged with [good first issue](https://github.com/mosaicml/composer/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) are great options to start contributing.
 
 If you have questions, join us on [Slack](https://join.slack.com/t/mosaicml-community/shared_invite/zt-w0tiddn9-WGTlRpfjcO9J5jyrMub1dg) -- we'll be happy to help you!
+
+We welcome contributions for bug fixes, new efficient methods you'd like to contribute to the community, or new models and datasets!
+
+## New Algorithms
+
+Have a new algorithm you'd like to contribute to the library as part of your research? Reach out to us on Slack or file a new issue with the proposed method, and we'll work with you to get started!
```
Comment (dblalock) [inline CONTRIBUTING.md:13]: What if people just built it for themselves and want to open a PR directly? "we'll work with you to get started" sounds 1) it's hard and 2) there's a hurdle to jump through, sort of like when a pricing page says "get in touch" instead of a price.

We can always reject PRs that aren't up to our standards, so I'd vote for saying new method PRs are welcome, but we suggest opening an issue first to discuss how and whether it makes sense to do it.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -5,3 +5,71 @@ Thanks for considering contributing to Composer!
...
+* `make test-dist-gpu WORLD_SIZE=2`  # run 2-gpu distributed tests
+
+These tests run with the `composer` launcher. We also support `WORLD_SIZE=1`, which would run the tests with the `composer` launcher on a single device.
+
+See the [Makefile](https://github.com/mosaicml/composer/blob/dev/Makefile) for more information.
+
+## Code Style & Typing
+
+Follow Google's
+[Python Style Guide](https://google.github.io/styleguide/pyguide.html) for how to format and structure code. Many of these are already taken care of by the `make style` command above.
```
Comment (dblalock) [inline CONTRIBUTING.md:72]: "these" -> "these guidelines", since "these" has unclear antecedent

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -5,3 +5,71 @@ Thanks for considering contributing to Composer!
...
+
+Composer aims to annotate all functions with type annotations (introduced in
+[PEP 526](https://www.python.org/dev/peps/pep-0526/). Not to worry if you are not a python typing expert, put in the pull request, and we'll help you with getting the code into shape.
```
Comment (dblalock) [inline CONTRIBUTING.md:75]:
```suggestion
[PEP 526](https://www.python.org/dev/peps/pep-0526/)). Don't worry if you are not a Python typing expert; put in the pull request and we'll help you get the code into shape.
```

Found this a little hard to parse because of the parens wrapping end of one sentence and then next sentence.

File: STYLE_GUIDE.md [hunk 12]
```diff
@@ -1,53 +1,45 @@
...
-Please do NOT address pylint errors in unrelated, unmodified files,
-as that will clutter the PR.
 
-
-## 2. Type Annotations and Typechecking
+## 3. Type Annotations and Typechecking
 
 Composer aims to annotate all functions with type annotations (introduced in
 [PEP 526](https://www.python.org/dev/peps/pep-0526/). Type annotations help statically catch `TypeError` and
 `AttributeError` bugs, in addition to other benefits, as outlined in the PEP.
```
Comment (dblalock) [inline STYLE_GUIDE.md:32]:
```suggestion
Composer aims to annotate all functions with type annotations (introduced in
[PEP 526](https://www.python.org/dev/peps/pep-0526/)). Type annotations help statically catch `TypeError` and
`AttributeError` bugs, in addition to other benefits, as outlined in the PEP.
```
Missing close paren.

File: STYLE_GUIDE.md [hunk 13]
```diff
@@ -1,53 +1,45 @@
...
 
-Composer uses [pyright](https://github.com/microsoft/pyright)
+Composer uses [PyRight](https://github.com/microsoft/pyright)
```
Comment (dblalock) [inline STYLE_GUIDE.md:34]: Pyright doesn't capitalize R. Also should probably just be lowercase to match usage below.

File: STYLE_GUIDE.md [hunk 14]
```diff
@@ -155,30 +142,67 @@ if parameter is None:
...
+        an optional dependency is missing.
+
+    1.  If the dependency is core to Composer, add the dependency to the `install_requires` section of
+        [setup.py](./setup.py) and the `requirements.run` section of [meta.yaml](./meta.yaml).
+
+### 6.2 Use of `__all__`
+
+All public modules must define `__all__` to be the list of members that should be re-exported.
+The variable is necessary to limit what `from XXX import *` imports, and so the documentation build
+only documents exported members (and not unrelated re-imports).
```
Comment (dblalock) [inline STYLE_GUIDE.md:181]:
```suggestion
The variable is necessary to 1) limit what `from XXX import *` imports, and 2) ensure that the documentation only includes exported members, not unrelated re-imports.
```
A little hard to parse. Expected "and so" to elaborate on for first clause, not introduce a second list item; and expected "build" to be a verb initially.

File: STYLE_GUIDE.md [hunk 15]
```diff
@@ -194,12 +218,112 @@ from composer.path.to.module import my_file as my_file
...
+        Args:
+            bar (int): Required bar.
+        """
+        ...
+
+    def foo2(bar: int = 42):
+        """Foo2.
+
+        Args:
+            bar (int, optional): Optional bar. (default: ``42``)
```
Comment (dblalock) [inline STYLE_GUIDE.md:244]: If we're going with
```...bar. Default: ``42`` ```
as the default value format, should show that here (and in below docstrings also)

Comment (dblalock) [reply STYLE_GUIDE.md:244]: Also I don't think we want people starting off with the word Optional in their variable descriptions (which the examples sort of suggest right now), so maybe change to something like:

```bar (int, optional): The first argument. Default: ``42`` ```

that can't as easily be read as suggesting particular free text content.

File: STYLE_GUIDE.md [hunk 16]
```diff
@@ -194,12 +218,112 @@ from composer.path.to.module import my_file as my_file
...
+    Any test fixtures for doctests should go in [docs/source/doctest_fixtures.py](docs/source//doctest_fixtures.py)
+    or in a `.. testsetup::` block.
+
+    For example:
+    ```python
+    def my_function(x: Optional[torch.Tensor]) -> torch.Tensor:
+        """blah function
+
+        Args:
+            input (torch.Tensor, optional): Your guess.
```
Comment (dblalock) [inline STYLE_GUIDE.md:301]: Can you do Optional args with no default value? maybe just remove the Optional here

**Review by dblalock — APPROVED** [review]
A few polishing suggestions but LGTM
